De-duplicate building `Spannable` on Android

cubuspl42 opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

A first step in my work on https://github.com/react-native-community/discussions-and-proposals/issues/695

De-duplicate the code for creating `Spannable` on Android. I'm planning to add quite serious new features to this module. This would be really hard with the current level of code duplication.

## Changelog:

[INTERNAL] [CHANGED] - De-duplicate building `Spannable` on Android

## Test Plan:

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

I tried to ensure that the refactored code is relatively easy to prove to be equivalent to the original duplicated one, but there's always a risk of a human mistake in this process. So far, I have been testing this by ensuring that nothing broke in the `Text` example section in RNTester.

cubuspl42 commented:
@NickGerleman I would be extremely grateful if you helped me with reviewing this or finding somebody to review this 😊

NickGerleman commented:
@mdvacca would you be able to take a look at this? It relates to removing duplication for Spannable generation.

IDK what our unit test coverage looks like here.

cubuspl42 commented:
@NickGerleman Thanks!

@mdvacca I would be grateful if you took a look at this! 🫶 

Unfortunately, if I were to answer the question, it would be "not so good"; I assume that if such unit tests existed, they would be located in `react-native/packages/react-native/ReactAndroid/src/androidTest`, and I don't think anything there is focused on testing `Spannable`.

cubuspl42 commented:
@NickGerleman @mdvacca 

A cat picture

![image](https://github.com/facebook/react-native/assets/2590174/fc650362-a0f8-463b-8085-b61b30b2475b)

cubuspl42 commented:
@NickGerleman @mdvacca

I can keep bumping. Let me know if there's anything I could do to move this forward

NickGerleman commented:
I've been checking with @mdvacca offline. I think he was sick last week, but mentioned he was planning to take a look and give some feedback. Thank you for keeping us honest here.

cubuspl42 commented:
@mdvacca It would be awesome if you could give it a quick look before the end of the week. I could start working on eventual changes before the next round.

mdvacca commented:
Hi @cubuspl42, thanks for the PR and apologies for the slow response. 
I will analyze in details and provide feedback tomorrow!

cubuspl42 commented:
@mdvacca
<img src="https://media2.giphy.com/media/FDmQjdAdaMb8wAD1VF/giphy.gif"/>

mdvacca reviewed (changes requested):
Hi @cubuspl42, thanks for working on this. It’s due for us to reduce the duplication of this code!

I wanted to share general feedback about the PR before getting into details:
- You are creating new public classes and interfaces. We recently created a new package called com.facebook.react.internal, where we are going to start moving public classes and interfaces that are part of the “core” but not part of the public API of React Native. I believe this is a good set of classes and interfaces that we can start including in there. Could you move these new classes and interfaces into “com.facebook.react.internal.view.text”?
- As part of the test plan you mention:
	> I tried to ensure that the refactored code is relatively easy to prove to be equivalent to the original duplicated one, but there's always a risk of a human mistake in this process. So far, I have been testing this by ensuring that nothing broke in the Text example section in RNTester.

Even if we have some tests for Text I don’t think they are enough for us to feel super comfortable about not breaking anything in production. The approach we usually take for these kind of refactors is running A/B test. From the PR standpoint, running an A/B test involves creating a feature flag in ReactFeatureFlags class and then refactor the PR to make sure we can gate the new changes. Can you add a new feature flag in ReactFeatureFlags and refactor your code to be able to A/B this refactor?

- You created this new class called “TextLayoutUtils” that contains several static methods that “add things to spannables or build spannables”. How about refactoring this code to use a builder pattern?
- All new public classes, interfaces and method must be documented
- OPTIONAL: We are starting to move the React Native codebase to Kotlin. This is not a must, but if you are very familiar with Kotlin it would be great to use kotlin in this refactor. Again this is not a hard requirement.

Thanks!

cubuspl42 commented:
@mdvacca 

I pushed a new revision, addressing most comments. What is left:

> How about refactoring this code to use a builder pattern?

Most usages of the builder pattern I've encountered were a substitute for the `data class` Kotlin `copy()` method; this isn't exactly a case like that.

Would you provide a few lines of code that show how things would look after the refactor so I'm sure we're on the same page?

> Can you add a new feature flag in ReactFeatureFlags and refactor your code to be able to A/B this refactor?

This is going to be tricky, but it is probably possible. 

From the operational perspective, you're talking about A/B tests that Meta could run on the Facebook app and/or other Meta apps, do I get this right?

cubuspl42 commented:
@mdvacca Let me know what you think of the Kotlin "port" of the new bits. For example, I used a named object for utils, instead of exposing public functions, as for now some of the names were quite generic, like `addText`. This might change once we close the other topic about the builder pattern.

cubuspl42 commented:
@mdvacca Bump

![IMG_1253_miniature](https://github.com/facebook/react-native/assets/2590174/39e9ba9f-6c33-4205-9f41-993418e2e26b)

cubuspl42 commented:
@mdvacca While I'm trying to take a humorous approach, this is a serious, externally sponsored effort, and we'll need multiple PRs to implement the desired functionality. I beg for some feedback loop fluency here.

I'll start figuring out the code for the feature flags, but feedback on rewriting everything to Kotlin will also be extremely helpful.

Please let me know if I, or Expensify, can do anything to improve the feedback loop situation.

mdvacca commented:
@cubuspl42, will take a look at it today, thank you!

cubuspl42 commented:
@mdvacca Thank you!

mdvacca commented:
@cubuspl42 

> From the operational perspective, you're talking about A/B tests that Meta could run on the Facebook app and/or other Meta apps, do I get this right?

Correct, technically this would mean to add a new static field variable in ReactFeatureFlags and then refactor the PR to be able to execute "old Code" or "new code" based on the value of the feature flag.

https://github.com/facebook/react-native/blob/97647d11841c9c97632617fe770439e3acb689ea/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/config/ReactFeatureFlags.java

We will take care of running the experiment at Meta and analyzing result of it.

I will share some detail of the code in a bit

mdvacca commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManager.java:
```diff
@@ -105,96 +104,9 @@ private static void buildSpannableFromFragment(
-              new SetSpanOperation(
-                  start, end, new CustomLineHeightSpan(textAttributes.getEffectiveLineHeight())));
-        }
+    final var textFragmentList = new BridgeTextFragmentList(fragments);
```
Let's say you create a feature flag called enableSpannableCreationUnification then here's one of the places where you have to gate your new code:

if (enableSpannableCreationUnification) {
   textFragmentList = new BridgeTextFragmentList(fragments);
   ...
} else {
  for (int i = 0, length = fragments.size(); i < length; i++) {
  ....
  }
}
```

I know this make this code a bit dirty, but it would be important to understand the impact of this PR

mdvacca commented:
Thanks for moving the new classes to internal and migrating them to Kotlin!

> Would you provide a few lines of code that show how things would look after the refactor so I'm sure we're on the same page?

I did a further analysis and I think that using a builder pattern will require a much bigger refactor, I think we could leave it like this for now.

Although it will important to add a FeatureFlag as I mentioned in my comment above.

Also, let's document all new public classes, interfaces and method using KDoc

@cortinico, can you also take a quick look at this PR? in particular the Kotlin side

Thanks!

cubuspl42 commented:
> Also, let's document all new public classes, interfaces **and methods** using KDoc

If I'm not mistaken, I added KDocs to all new classes. Are you sure about the methods? Should I add comment docs for all Java getters, Kotlin properties, the new `addFoo` helpers, etc.?

cubuspl42 commented:
@mdvacca I rebased on latest `main` and added the feature flag. I also ensured that the diff with `main` is minimal.

cubuspl42 commented:
Also converted some left-over JavaDocs style links to KDoc-style (oops!)

cubuspl42 commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/internal/views/text/BasicTextAttributeProvider.kt:
```diff
@@ -0,0 +1,41 @@
+import com.facebook.react.uimanager.ReactAccessibilityDelegate
+
+/**
+ * Interface for an entity providing basic text attributes of a text node/fragment. "Basic" means
```
"Basic" can be renamed to "Direct" if you find that better suiting

cubuspl42 replied:
It's more of a technical thing than semantical

cubuspl42 commented:
> We will take care of running the experiment at Meta and analyzing result of it.

How long, **approximately**, would this take? Does it block the next steps of my project?

mdvacca commented:
> How long, approximately, would this take? Does it block the next steps of my project?

The experiment shouldn't block you to next steps for the project.
Thanks for updating the code, will take a look today

cubuspl42 commented:
@mdvacca Thank you for this info! I see that there's a conflict, but it's probably minor; I'll resolve it soon

cubuspl42 commented:
@mdvacca Rebased on `main` and fixed a minor conflict

cortinico commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/internal/views/text/fragments/TextFragmentList.kt:
```diff
@@ -0,0 +1,10 @@
+package com.facebook.react.internal.views.text.fragments
+
+/**
+ * Interface for a list of text fragments
```
```suggestion
 * Interface for a list of [TextFragment]
```

cubuspl42 replied:
I changed this and a few other comments 👍

cortinico commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/internal/views/text/fragments/TextFragment.kt:
```diff
@@ -0,0 +1,24 @@
+interface TextFragment {
+  fun getTextAttributeProps(): TextAttributeProps
+
+  fun getString(): String?
```
Can those and other become properties of this interface?

cubuspl42 replied:
Sure, I missed it. Done.

cortinico commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/internal/views/text/fragments/MapBufferTextFragmentList.kt:
```diff
@@ -0,0 +1,14 @@
+/**
+ * A list of text fragments backed by a [MapBuffer]
+ */
+class MapBufferTextFragmentList(private val mFragments: MapBuffer) : TextFragmentList {
```
```suggestion
class MapBufferTextFragmentList(private val fragments: MapBuffer) : TextFragmentList {
```

cubuspl42 replied:
I was just following the existing RN conventions, but I assume that when switching to Kotlin they can change too

cubuspl42 replied:
...but it seems I wasn't consistent within this PR

cortinico commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/internal/views/text/fragments/BridgeTextFragmentList.kt:
```diff
@@ -0,0 +1,14 @@
+/**
+ * A list of text fragments backed by a [ReadableArray]
+ */
+class BridgeTextFragmentList(private val mFragments: ReadableArray) : TextFragmentList {
```
```suggestion
class BridgeTextFragmentList(private val fragments: ReadableArray) : TextFragmentList {
```

cortinico commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/internal/views/text/fragments/BridgeTextFragment.kt:
```diff
@@ -0,0 +1,28 @@
+/**
+ * A text fragment backed by a a [ReadableMap]
+ */
+class BridgeTextFragment(private val mFragment: ReadableMap) : TextFragment {
```
```suggestion
class BridgeTextFragment(private val fragment: ReadableMap) : TextFragment {
```

cortinico commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/internal/views/text/EffectiveTextAttributeProvider.kt:
```diff
@@ -0,0 +1,24 @@
+ * Interface for an entity providing effective text attributes of a text node/fragment
+ */
+interface EffectiveTextAttributeProvider : BasicTextAttributeProvider {
+  companion object {
+    const val UNSET = ReactFontManager.TypefaceStyle.UNSET
+  }
```
Do we need this redefined here? Why?

cubuspl42 replied:
Yeah, it's smelly... `TypefaceStyle` is a tuple in the form `TypefaceStyle(int weight, boolean italic)`, while I'm using this for font size. So not completely the same thing. So I kind of "stole" their constant...
 
Possible ways forward here:

- Switching to `Int?` for `effectiveFontSize` (with some possible minor performance penalty)
- Redeclaring `UNSET = -1` in this interface
- Refactoring these two constants into something like `FontConstants.UNSET`

cubuspl42 replied:
I'd prefer the `FontConstants.UNSET` path, as it doesn't introduce inconsistency with the existing code and feels acceptably clean.

cubuspl42 replied:
You can quickly search for `UNSET =` in the repository. In a few places, we define it on a per-class basis. It's typically `-1`, but also a minimal `Long` value in one place.

cubuspl42 replied:
The issue starts here:

https://github.com/facebook/react-native/blob/285629922cafd86914af9ed95ff1fbbd2c5a7c79/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextShadowNode.java#L83

`ReactTextShadowNode` uses the constant from `ReactBaseTextShadowNode`...

https://github.com/facebook/react-native/blob/285629922cafd86914af9ed95ff1fbbd2c5a7c79/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactBaseTextShadowNode.java#L58

...which in turn "steals" it from `TypefaceStyle`.

So this is what I originally replicated and you asked about.

cubuspl42 replied:
So, actually, `FontConstants` is also not a good name/source for this constant. In my latest revision, I suggest an approach where all files I create/modify depend on the `TextAttributeProps.UNSET` static property.

This reduces the total number of `UNSET = -1` definitions in the repository to three.

cortinico reviewed (commented):
I've done a pass on the Kotlin side, looks great mostly nits!

mdvacca commented:
importing PR to check build

mdvacca reviewed (approved):
lgtm, thanks for working on it.

cubuspl42 commented:
@mdvacca Awesome, thanks!

The linter is failing, but I don't think I have access to the error logs; let me know if you can copy-paste them here

mdvacca commented:
@cubuspl42, We have to make some changes internally before landing it, we couldn't make it this week, we will work on them next week and land before end of the week.
we will fix the linter before we land the PR.

CC @NickGerleman

cortinico commented:
> @mdvacca @NickGerleman

Hey @cubuspl42, David is off. Please don't ping him further. @NickGerleman is going to take this over but is also quite overloaded at the moment so please allow some time

NickGerleman commented:
Sorry for the late response here. It is okay to give me a reminder, but as cortinco mentioned, my focus has been divided over the last couple of weeks (it didn't help I was oncall for one of them).

It sounds like @mdvacca thought this was ready to ship, and I can see that we added a feature flag, which is especially helpful given Meta is going into a code freeze season, encouraging the ability to turn new larger changes off.

There are some errors when we tried to build this using internal Buck build. I rebased to trigger a new Gradle build, to double check how they might be acting differently.

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactBaseTextShadowNode.java:
```diff
@@ -221,6 +259,93 @@ private static void buildSpannedFromShadowNode(
+            supportsInlineViews,
+            inlineViews,
+            sb.length());
+      } else if (child instanceof ReactTextInlineImageShadowNode inlineImageChild) {
```
```suggestion
      } else if (child instanceof ReactTextInlineImageShadowNode) {
```

Formatter, but oddly not compiler, caught invalid token here

NickGerleman replied:
Got it. It looks like the version of `google-java-format` we use for formatting is old enough to not be able to handle the new syntax. Could we use something more compatible?

NickGerleman commented:
Okay, I took a look at internal build failures. The TLDR: is that dependencies are handled more granularly in internal build (similar granularity to a target per package). This disallows dependency cycles between packages, where this change would introduce a couple.

1. `com.facebook.react.common` must now depend on `com.facebook.react.views.text`. We can fix that pretty easily by reverting the change in `ReactFontManager.java`
2. `com.facebook.react.internal.views.text` and `com.facebook.react.views.text` both depend on each other. I think that would need a larger change to fix.

There might be more after those are unpeeled.

cubuspl42 commented:
> 2\. I think that would need a larger change to fix.

It's not realistic to do what these classes aim to do with this requirement to keep Java packages acyclic.

These classes landed in the `internal` package because that was the PR review feedback.

The `internal` package is quite thin, and it didn't have any "mirrored" packages like `view` / `view.text` before. Maybe we now know the reason?

Could we agree that I move back the new classes to `react.views.*`? I could mark them "internal" with some kind of annotation or appropriate comment, so it's clear that they are not considered stable public APIs.

I think that you folks at Meta could figure out internally how to handle this conflict of requirements (i.e. "let's put implementation detail parts in `internal.*`" vs "let's not have package cycles"), because as an external contributor, my hands are tied.

cubuspl42 commented:
@NickGerleman 

I...
- rebased the branch on top of upstream `main`
- cleaned up the commit structure a bit
- undone the `internal.view` mirrored package
- extracted `UNSET` to a shared constant in `common` (so it doesn't create cycles)

Would you, please, take a look?

NickGerleman commented:
I will try importing and running by the internal builds/tools again. @mdvacca should also be back from leave, and now holidays, and I let him know as well.

NickGerleman commented:
That went a lot more smoothly. There were a couple small changes I made to let it build internally.

1. Formatting/copyright licenses for the lint tools
2. Internal build doesn't like Java 10 `var`
3. Several Kotlin interfaces were marked as `internal`, but the internal build has modules at a smaller granularity, forcing them to be changed to public.

I'm exporting the change back, for review, but otherwise, I think the final step is getting a stamp from @mdvacca with the new changes.

I would bias towards getting this merged, and doing any sort of organizational followup later.

cubuspl42 commented:
> 3\. Several Kotlin interfaces were marked as `internal`,

I did that as a compromise in the context of the PR comments. Moving things to `react.internal.views` didn't work, so I wanted to provide something in exchange.

It's not my personal need or opinion, un-marking these as `internal` is totally fine for me.

> Internal build doesn't like Java 10 var

Hm, maybe that tool is just stuck on Java 8. I will keep that in mind.

> I would bias towards getting this merged, and doing any sort of organizational followup later.

I'm not sure which followups you mean precisely, but I'm also biased toward getting this merged ❤️

cubuspl42 commented:
@NickGerleman 

Do I understand correctly that no action is required from me, for now?

cubuspl42 commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextAttributeProps.java:
```diff
@@ -84,16 +84,17 @@ public class TextAttributeProps {
 
-  // `UNSET` is -1 and is the same as `LayoutDirection.UNDEFINED` but the symbol isn't available.
-  protected int mLayoutDirection = UNSET;
+  // `Constants.UNSET` is -1 and is the same as `LayoutDirection.UNDEFINED` but the symbol isn't available.
```
Ouch, this was unintended

cubuspl42 replied:
Fixed; I also updated the comment slightly

cubuspl42 commented:
@NickGerleman @mdvacca Friendly bump. I'm sorry to bother you, but if I don't keep pinging, this will never be merged.

NickGerleman commented:
Let me ping @mdvacca again internally about this. Sorry about the delays here, I agree we dropped the ball.

cubuspl42 commented:
@mdvacca Would you take a look, **please**?

The code is nearly the same as when you last looked at it. We have a feature flag, and based on it, we switch over unified and duplicated implementations. The most important aspect of the review is that you confirm that the old behavior is unchanged when the flag is off; but it's quite clear from the diff.

Another minor change is extracting the `-1` integer value to a separate constant, treating it somewhat analogically to `Float.NaN` or `null` (the convention was already a fact in the codebase, now it's formalized).

This is basically it. I'd say it can be reviewed in a few minutes.

mdvacca reviewed (approved):
lgtm, there are few follow up that @NickGerleman will work on internally before merging.
We are planning to merge this PR early Thursday next week.

cubuspl42 commented:
@NickGerleman @mdvacca The big day!

cubuspl42 commented:
Did we miss the launch window? ☹️

NickGerleman commented:
@cubuspl42 I wasn't able to get to this yesterday, but am working on it now (importing latest revision, then re-applying the changes mentioned above).

@mdvacca also mentioned merging IntConstants and ReactConstants, which I will see if I can just inline on top of your change to short-circuit the back and forward.

Apart from setting up the experiment, which I can take on, I'm also going to quickly run this, enabled in code, against a fuller set of product UI tests. This doesn't normally happen when a flag is only enabled by experiment, but these tests can help us find visual differences nice and early (experiments are more helpful in finding crashes, perf issues, changes which impact user behavior, and making it so we can turn the change off it it causes any issues in a release out in the wild).

NickGerleman commented:
@cubuspl42 this landed 40 minutes ago. There is a lag in propagating to GitHub, but the change is there.

cubuspl42 commented:
@NickGerleman

<img src="https://media4.giphy.com/media/mi6DsSSNKDbUY/giphy-downsized-medium.gif"/>

Awesome! Thank you so much! 🚀

cubuspl42 commented:
> Apart from setting up the experiment, which I can take on

What is the status of this? It would be great to know the ETA for this experiment and its end, however long it might be.

cubuspl42 commented:
@Kudo Sorry to hear that! 🙁

The deal is that it's very unclear which parts of React Native `ReactAndroid` Java APIs are actually "public". There's an ongoing effort to extract stuff to an `internal` package to make it more clear, but it also has its problems as we discovered in this PR.

I can add back `TextAttributeProps.UNSET` as an "alias" for the new `UNSET` constant.

Kudo commented:
no worries! i think meta started an effort to define public apis for react-native.
i would just patch lottie-react-native for [replacing `UNSET to -1`](https://github.com/expo/expo/pull/26587/files#diff-cecabf9e2f8629fb8c1af7cc2b1d772fdd7fab6edaef14d52c59796fc561bbecR1), and let you folks to follow up a better solution for compaibility issue.

TheRogue76 commented:
It would be nice, we are also exploring doing a V7 for lottie to address this and the new install_module script on iOS (since it only supports RN 70 and above). But it would be nice to not have to worry about this too much 🙏🏻

cubuspl42 commented:
@Kudo 

> i would just patch lottie-react-native for [replacing UNSET to -1](https://github.com/expo/expo/pull/26587/files#diff-cecabf9e2f8629fb8c1af7cc2b1d772fdd7fab6edaef14d52c59796fc561bbecR1)

I'm not an authoritative source here, but I'm 90% sure this is now considered an public API:

https://github.com/facebook/react-native/blob/7bcdb23cf06997e50fee7cca41735b77881ee144/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/common/ReactConstants.java#L20

I'd use this on your place as a fix for the specific problem you encountered.

Kudo commented:
@cubuspl42 sounds good! i'll update the patch accordingly. thanks for the great hint.

cubuspl42 commented:
For future reference, there's one other local definition of `UNSET`:

https://github.com/facebook/react-native/blob/714b502b0c7a5f897432dbad388c02d3b75b4689/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/TouchEvent.java#L33

If it ever needs extracting, we could name it `ReactConstants.UNSET_LONG_MIN`. However, it might never be necessary.

I don't think we have any cases of `UNSET = -1L`. If we need it, we could name it `ReactConstants.UNSET_LONG`.

I think we don't need more future predicting here.

cubuspl42 commented:
@NickGerleman Kind bump on the experiment info

What's the ETA for starting and finishing it?

NickGerleman commented:
> @NickGerleman Kind bump on the experiment info
> 
> What's the ETA for starting and finishing it?

Because of the nature of needing to incrementally roll out to production audience, and get feedback, it will be a while before we are able to remove the forking (potentially a month or more). But it should be possible to continue building on top of it in the meantime.

cubuspl42 commented:
@NickGerleman Thanks!

As I understand it, it's already live, and the finishing ETA is roughly 1-3 months. I don't mean to push it, I just want to establish a timeline.

NickGerleman commented:
Hey @cubuspl42, while rolling this out to our users, we discovered it causes a top-line performance regression, in the time to render content, for a couple of our scenarios. More concretely, a number of users in the millions would go from "responsive" to "unresponsive" on those surfaces.

I have deallocated the experiment due to performance impact. If we wanted to try shipping this again, I think we'd need to understand where this issue presents in real apps, and come up with a fix.

cubuspl42 commented:
Thank you for sharing these results! I must admit that they surprised me, but prove you had a good intuition to set up an experiment. Did the mentioned user group share a common configuration of React Native? Was it Fabric/bridgeless?

NickGerleman commented:
The surfaces were Fabric + Bridgeless. One involved a scrollable grid of components with text. One was more static. Both in an app with very large usage.

A lot of real-world Android devices are low/middle-end, which means sometimes a very minor change on higher end device, can push a surprisingly large number of real-world devices from one responsiveness threshold to another.

cubuspl42 commented:
It was brought to my attention that someone from Meta (I'm not sure whether it was you personally) contacted Expensify, informing that Meta will no longer accept big PRs related to `Text` in the foreseeable future.

1. Would you like me to submit a follow-up that removes the "unified" path, or would you prefer to leave it there, so someone from Meta can investigate this topic in the future, when this will be more aligned with Meta's roadmap and priorities?

2. Other folks from Expensify are wondering whether adding test suites of some kind could change Meta's position on this matter. If you know that the answer is "definitely no," I could forward this information, and it would speed up the pivoting process on our side (possibly by months).

Thank you for your help so far; I'm sorry it ended this way.

NickGerleman commented:
Hey there. I think @TheSavior chatted with folks at Expensify wrt Text changes.

Our own problem right now, is that certain classes of contributions are hard to land, from outside of Meta. In these cases, where we have encouraged contribution, we've often created a bad experience for change authors (hard/slow to merge, hard to debug regressions effecting product code), coupled with at times a spikey workload for engineers at Meta, who usually don't have time budgeted ahead of time for larger changes outside their current area of focus.

I don't know that we ever explicitly chatted about what to do with in-flight changes. If there is a scoped, logical wrap-up, to changes already in flight, I think we could take a look at those. But if we think this is more likely to become dead code, I do really appreciate the offer to help clean things up.

I really appreciate all the time you spent here, working to make RN better. I am sorry, we couldn't offer a better experience here.

TheSavior commented:
I do think it is a reasonable question of whether there is something that could be contributed around testing to increase our confidence in merging changes to Text. @NickGerleman do you have any thoughts there?

NickGerleman commented:
Meta has collectively spent some tens of thousand of hours on tooling and infra, but has usually focused on using them for Meta products, or the Meta monorepo. That puts us into a complicated state, where we do use this tooling, as a pragmatic way to ensure RN quality (for our apps, and OSS). But those tools don't come free to OSS.

Without commenting about what can be contributed vs not, there are a few classes of tooling that come to mind, that we use to ensure quality:

### Fast and stable screenshot comparison testing

We get feedback on whether a change visually effects both our real-world surfaces, and fixtures we created. E.g. when I was working on the Samsung TextInput SEV a while back, I set up infra to visually compare a lot of interesting formatting span/nesting scenarios, for both Fabric and Paper (and they have since caught real regressions).

There have been some efforts to stand up OSS E2E infra before, but they have run into stability issues, and never got around to a lot of what is really useful (e.g. screenshot comparison). If we had reliable, simple infra, to make it easy for contributors to add image baselines, that would give us a lot more confidence for some changes.

### Experimentation infra

We often test by observing differences in behavior/performance when a change is enabled vs not. This is inherently something that has to come from a collection of product code (e.g. when I was at another company, we did the same thing). That looks like, enabling a change for a large fleet of users, across surfaces, and having good enough metrics to understand if we see e.g. more rage shakes, crashes, perf changes, impression changes, etc.

### Perf infra

I know things like the Java sampling profiler work pretty out of the box in OSS (and I think, that is probably the best place to start looking for diagnosing a perf issue for e.g. a change like this). But there is often bespoke infa Meta has to make this sort of thing easier and faster. E.g. If I see that a change makes a surface slower to start, I can go look at how much time is spent, via a sampling profiler, that ran during the duration, collecting information across real production users. Or, tooling for e.g. trace markers, which I'm not sure are wired to sinks in most OSS builds (I remember when I was at MSFT, we did wire them to Windows ETW).

NickGerleman commented:
To clarify though, I don't think better tooling is a magic bullet to ensure smooth contributions, though for many cases, I think it does help build confidence. I think another component, is aligning changes which will require a lot of interaction from engineers at Meta, to work the engineer is chartered to. That helps make it a lot easier to prioritize the work, and for someone to represent it or deal with any fallout when shipping it. I can say personally, at any given moment, there will be something falling off my plate that someone has asked me to do, just by the nature of the scale factor of the relatively small number of folks working on RN, compared to the very large userbase. So, it's easier to keep things above that line, when we can very easily justify the work.

cubuspl42 commented:
These are extremely helpful thoughts; this is a communication breakthrough from my perspective.

I hope I'm not stepping out of the line here, but this has been bothering me:

Would it be possible for you to talk with your superiors and discuss Meta's priorities in the context of React Native within its [Open Source](https://opensource.fb.com/) umbrella project? Would be possible to establish a budget for this "communitization" effort of React Native? A few talented engineers from the Expensify project (and the supporting companies, like Software Mansion) would love and be able to help with this, but I'm not going to lie; these are very difficult technical tasks requiring expert knowledge, experience **and time**.

The budget established for implementing spans (eventually, inline code blocks) in `Text` wouldn't cover implementing the testing infrastructure prototype. Also, it may not directly fall into a category of efforts that Expensify could manage to sponsor in general. The gains from these efforts would benefit the whole community, though.

cubuspl42 commented:
Let's check if [these tenets](https://opensource.fb.com/) are more than marketing 😉

> Meta is committed to the notion that open source not only provides great technology for developers, but also brings the best out in people. Meta believes in the following tenets:
>
> #### Collaboration
> 
> Open source allows developers from around the world to proactively work together to achieve common goals.
>
> #### Community
> 
> Open source provides opportunities for the creation of diverse communities in support of the betterment of everyone.
>
> #### Technology
> 
> Open source has produced historic technologies, including operating systems, frameworks and programming languages.

fabOnReact commented:
The facebook/react-native Android Test Suite was disabled  for components that rely on Yoga for layout (commit https://github.com/facebook/react-native/commit/709a441ecf54cd9465f5946af0454ee7d10d5cbe).
It could be solved using the mock [YogaConfigFactory](https://github.com/facebook/yoga/blob/996267dbcb1cd809f19f8d8e47d075bb982e32a8/java/tests/com/facebook/yoga/YGFlexDirectionTest.java#L29) to run those tests with roboeletric. 

I don't believe Yoga layout is essential to run tests for Text and TextInput. 
- We could just mock the height/width of the EditText or TextView. No need to compute height/width with Yoga for the test case.
- The Text/TextInput spans do not use Yoga Layout. It is an Android API named [Spans](https://developer.android.com/develop/ui/views/text-and-emoji/spans#java). We don't need to have Yoga to verify that they work correctly.

More info in PR https://github.com/facebook/react-native/pull/35949 and more useful code pointers in comment https://github.com/facebook/react-native/pull/35130#issuecomment-1423867378.
